src/Control/Distributed/Spark/Closure.hs

```diff
@@ -22,27 +22,35 @@ module Control.Distributed.Spark.Closure
 import Control.Distributed.Closure
 import Control.Distributed.Closure.TH
 import Data.Binary (encode, decode)
+import qualified Data.Coerce as Coerce
 import qualified Data.ByteString.Lazy as LBS
 import Data.ByteString (ByteString)
 import Data.Typeable (Typeable)
+import Foreign.ForeignPtr (newForeignPtr_)
+import Foreign.ForeignPtr.Unsafe (unsafeForeignPtrToPtr)
 import Foreign.JNI
+import Foreign.Ptr (Ptr)
 import Language.Java
 
 -- | The main entry point for Java code to apply a Haskell 'Closure'. This
 -- function is foreign exported.
+--
+-- The function in the closure pointed by the first argument must yield
+-- a local reference to a Java object, or the reference might be released
+-- prematurely.
 apply
-  :: JByteArray
-  -> JObjectArray
-  -> IO JObject
+  :: Ptr JByteArray
```

Comment:
This change in type signature makes it look like you're adding an extra level of indirection? But are you? It looks like you can simply unwrap `J`.